src/idpf.rs

```diff
@@ -0,0 +1,1869 @@
+//! This module implements the incremental distributed point function (IDPF) described in
+//! [[draft-irtf-cfrg-vdaf-03]].
+//!
+//! [draft-irtf-cfrg-vdaf-03]: https://datatracker.ietf.org/doc/draft-irtf-cfrg-vdaf/03/
+
+use crate::{
+    codec::{CodecError, Decode, Encode, ParameterizedDecode},
+    field::{Field255, Field64, FieldElement},
+    prng::Prng,
+    vdaf::{
+        prg::{Prg, RandSource, Seed, SeedStream},
+        VdafError, VERSION,
+    },
+};
+use bitvec::{bitvec, boxed::BitBox, prelude::Lsb0, slice::BitSlice, vec::BitVec};
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt::Debug,
+    io::{Cursor, Read},
+    ops::Index,
+};
+use subtle::{Choice, ConstantTimeEq};
+
+/// IDPF-related errors.
+#[derive(Debug, thiserror::Error)]
+pub enum IdpfError {
+    /// Error from incompatible shares at different levels.
+    #[error("tried to merge shares from incompatible levels")]
+    MismatchedLevel,
+
+    /// Invalid parameter.
+    #[error("invalid parameter: {0}")]
+    InvalidParameter(String),
+}
+
+/// An index used as the input to an IDPF evaluation.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct IdpfInput {
+    /// The index as a boxed bit slice.
+    index: BitBox,
+}
+
+impl IdpfInput {
+    /// Convert a slice of bytes into an IDPF input, where the bits of each byte are processed in
+    /// LSB-to-MSB order. (and subsequent bytes are processed in their natural order)
+    pub fn from_bytes(bytes: &[u8]) -> IdpfInput {
+        let bit_box_u8_storage = BitBox::<u8, Lsb0>::from_boxed_slice(Box::from(bytes));
+        bit_box_u8_storage.as_bitslice().into()
+    }
+
+    /// Convert a slice of booleans into an IDPF input.
+    pub fn from_bools(bools: &[bool]) -> IdpfInput {
```

Comment:
I mainly wanted to provide this as an entry point using `std` types only, so that there was a way to construct inputs without dealing with `bitvec` types. I added `clone_with_suffix()` as a helper for the candidate prefixes case.